JD-4-5-6/src/main/java/com/example/weekone/WeekOneApplication.java

```diff
@@ -25,9 +24,143 @@ public static void main(String[] args) {
+        for (int i = 1; i < fibonnaciList.size(); i++) {
+            // Previous index is the current index value - 1
+            int previous = i-1;
+            // Current index integer value
+            Integer current = fibonnaciList.get(i); //key
+            // While previous value is greater than the current value
+            while ((previous > -1) && (fibonnaciList.get(previous) > current)) {
+                // Set the values
+                fibonnaciList.set(previous+1, fibonnaciList.get(previous));
+                previous--;
+                whileLoopCount++;
+            }
+            fibonnaciList.set(previous+1, current);
+            outerLoopCount++;
+        }
+
+        System.out.println("Outer Loop count: " + outerLoopCount);
+        System.out.println("While Loop count: " + whileLoopCount);
+        return fibonnaciList;
+    }
+
+    /**
+     * Bubble Sort
+     * Definition: Sort parameter List with Integer 233 added at the 0 index of FibonacciSequence
+     * @param fibonnaciList
+     */
+    public static void bubbleSort(List<Integer> fibonnaciList){
+        // Loop count to count the amount of loops this BubbleSort took
+        int outerLoopCount = 0;
+        int innerLoopCount = 0;
+
+        fibonnaciList.add(0, 233);
+        int n = fibonnaciList.size();
+        // Loop through the length of the collection list
+        for (int i = 0; i < n-1; i++) {
+            // For Each element loop through and compare
+            for (int j = 0; j < n - i - 1; j++) {
+                innerLoopCount++;
+                // If the current value is greater than the next only switch
+                if (fibonnaciList.get(j) > fibonnaciList.get(j + 1)) {
+                    // Current value to hold as a variable
+                    int current = fibonnaciList.get(j);
+                    // Next value to hold as a variable
+                    int next = fibonnaciList.get(j + 1);
+                    // Set the New current and the next values, so the current is now less than the next
+                    fibonnaciList.set(j + 1, current);
+                    fibonnaciList.set(j, next);
+                }
+            }
+
+            outerLoopCount++;
+        }
+        System.out.println("Outer Loop count: " + outerLoopCount);
+        System.out.println("Inner Loop count: " + innerLoopCount);
+    }
+
+    /**
+     * Selection Sort
+     * Definition: Sort parameter List with Integer 233 added at the 0 index of FibonacciSequence
+     * @param fibonnaciList
+     *
+     * Notes: Outer Loop count: 12, Inner Loop count: 78, Total count = 90 loops.
+     */
+    public static void selectionSort(List<Integer> fibonnaciList){
+
+        // Loop count to count the amount of loops this InsertionSort took
+        int outerLoopCount = 0;
+        int innerLoopCount = 0;
+
+        // Initialize with 233 at index of 0 to provided array
+        fibonnaciList.add(0, 233);
```

Comment:
Since you kept everything as a Java collections api list what happens under the hood when you insert an element into a specific location to the other element already in index zero?

Reply:
Extra Extra Credit questions (dont stress yourself if you dont know the answers, entirely for fun. "I dont know" is an acceptable answer 😄 ) 

• What is the run time complexity (Big O notation) of modifying a List vs an Array? 
• Whats the difference between a List, an Array, and (for the curve ball) ArrayList? 
• What are the cost/benefits of using `Integer` vs `int` ?

Reply:
If you had instead implemented this using a simple array, what would have to happen in order to insert 233 at index 0? Is that different from a List? If so, why?

Reply:
Arrays are faster in their execution, but they can consume more memory. It only allows the amount that is set unless we make a copy and assign a new array in memory with the updated data length. Arrays also only allow the same data type, but that is part of their overall performance efficiency. 

Collections are dynamic, and take less memory. They don't perform as well, but it is a small nano second difference. 

Question 1: When I insert an element into a specific location it shifts everything to the right. The worst case scenario would be to shift everything to the right at index of 0, and the best case to add an element at the end of the list. 

Question 2: Inserting a value into a List vs an Array? My thoughts with this go: Inserting into an array would be O(n) if we are having to copy and insert with an array fixed length. If we are re assigning an index array value would be O(1). 
With a List I would put this the same? O(n) to add to a List, so each value can shift over. ?? Going off my gut thought, so if I'm wrong please correct me LOL)

Question 3: A List is a child interface that extends Collection (Inherits the parent class of Collection to the child interface of List). -> An ArrayList is class that implements the List interface, but the difference is An Array list is a dynamic array that contains objects, and a List is used to create a List of objects that are associated with their index numbers. List is a Interface, ArrayList is a Class.

Question 4: Benefits/Differences of int and Integer. Both are stored as the same 32 bit value for computer memory. int can provide less flexibility. An Integer (object) can be used and manipulated with data. You can't int.parseInt("10"), but you can Integer.parseInt("10").

Reply:
> Collections are dynamic, and take less memory. They don't perform as well, but it is a small nano second difference.

Eh, Id actually argue the opposite. Collections tend to be More memory heavy because they are doing a lot of extra 'stuff' under the hood. Arrays and some of the other more primitive types of data structures tend to be lightweight and quick. However, they lack flexibility and are fairly restrictive (you cant expand the size of an Array for instance, you need to create a whole new array). To your point, collections arent always the quickest, but what you get in terms of convenience far outweighs the loss in performance. That said, both are valuable tools in the tool belt, knowing the tradeoffs of both is in your best interest :) 

> List is a Interface, ArrayList is a Class

Bingo. So the nuance here is that an ArrayList implements the List Interface, but uses Arrays under the hood as part of its implementation. Thats where it gets its speed from. 

> Question 2: Inserting a value into a List vs an Array? My thoughts with this go: Inserting into an array would be O(n) if we are having to copy and insert with an array fixed length. If we are re assigning an index array value would be O(1).
With a List I would put this the same? O(n) to add to a List, so each value can shift over. ?? Going off my gut thought, so if I'm wrong please correct me LOL)

So what you described in overwriting a value in an Array is basically access, which as you correctly pointed out, is O(1) or constant time. Arrays are very good at that since all we need is the address in memory to get the value. 

For Lists, its a little more interesting. Accessing a List is O(n), because you need to traverse the List to reach the element of interest. However, Insertion is actually really fast O(1). A List is typically a  LinkedList implementation, which stores a reference to its neighbor. When you need to do the actual insertion its a simple matter of pointing the references to the right objects (its a common 1st year CS assignment, easy in concept but the devil is in the details).

 >Question 4: Benefits/Differences of int and Integer. Both are stored as the same 32 bit value for computer memory. int can provide less flexibility. An Integer (object) can be used and manipulated with data. You can't int.parseInt("10"), but you can Integer.parseInt("10").

This is actually similar to the difference between an Array and ArrayList. Like and Array, and `int` is a primitive type and very lightweight but largely restrictive in its use. However, what you loose in flexibility, you gain in speed and memory cost. By contrast, an `Integer` like an ArrayList is a much heavier object in terms of memory usage. However, as outlined, an ArrayList is dynamic and very flexible, and the same is true with an `Integer` object. The biggest (though by no means only) advantage is nullability. You arent going to get an error because you tried to assign Null to your Integer object. 

 Wont comment on the bit value, but the main difference between a primitive `int` compared to the object `Integer` is that the object can be nullable, while an `int` cannot. The primitive is also much faster

Reply:
Big O is a lot more nuanced than this, but Ive found this to be a great cheatsheet when needing to think about Big O https://www.bigocheatsheet.com/  (I actually have a poster of this in my office)